includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -922,6 +923,25 @@ private function validate_cdata_for_node( DOMElement $element, $cdata_spec ) {
 				return [ 'code' => self::MANDATORY_CDATA_MISSING_OR_INCORRECT ];
 			}
 		}
+
+		// When the CDATA is expected to be JSON, ensure it's valid JSON.
+		if ( 'script' === $element->tagName && in_array( $element->getAttribute( 'type' ), [ 'application/json', 'application/ld+json' ], true ) ) {
+			if ( empty( $element->textContent ) ) {
+				return [
+					'code'    => self::INVALID_JSON_CDATA,
+					'message' => esc_html__( 'Expected JSON, got an empty value', 'amp' ),
+				];
+			}
+
+			json_decode( $element->textContent );
+			if ( JSON_ERROR_NONE !== json_last_error() ) {
+				return [
+					'code'    => self::INVALID_JSON_CDATA,
```

Comment:
Please also include the JSON error code (perhaps `json_error`). Ultimately, this may be preferable than using the `message` as it would mean the strings could be translated, e.g. using https://www.php.net/manual/en/function.json-last-error-msg.php#117393

Reply:
Or rather, each of the error codes could have their own dedicated validation error codes. So instead of just `INVALID_JSON_DATA` there could be one for each of the constants in https://www.php.net/manual/en/function.json-last-error.php

Then the specific error message could be displayed at runtime rather than being stored in the actual error object, allowing for translations tailored to the user.

Reply:
Good point, these error codes could be from the constants that `json_last_error()` returns. I'll apply that later tonight if that's alright.

Reply:
>Then the specific error message could be displayed at runtime rather than being stored in the actual error object, allowing for translations tailored to the user.

Sure, it'd be good to have actual translations of these error messages.

But do you have in mind passing each of the [10 error messages](https://www.php.net/manual/en/function.json-last-error.php#refsect1-function.json-last-error-returnvalues) through `__()`, like:

```php
private static function get_error_message() {
	...
	case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_DEPTH:
		return __( 'The maximum stack depth has been exceeded', 'amp' );
	case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_STATE_MISMATCH:
		return __( 'Invalid or malformed JSON', 'amp' );
	etc...
```

This could get pretty verbose, but maybe I'm not understanding this.

Thanks, Weston!

Reply:
Yes, something like that. Maybe let the error title be normalized to just “Invalid JSON” for each error code, but then have the message have more elaboration, for example:

```diff
diff --git a/includes/validation/class-amp-validation-error-taxonomy.php b/includes/validation/class-amp-validation-error-taxonomy.php
index 6a6158c61..ed6c49e40 100644
--- a/includes/validation/class-amp-validation-error-taxonomy.php
+++ b/includes/validation/class-amp-validation-error-taxonomy.php
@@ -1833,8 +1833,22 @@ class AMP_Validation_Error_Taxonomy {
 					$content .= '</p>';
 				}
 
-				if ( isset( $validation_error['message'] ) ) {
-					$content .= sprintf( '<p>%s</p>', esc_html( $validation_error['message'] ) );
+				$message = null;
+				switch ( $validation_error['code'] ) {
+					case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_DEPTH:
+						$message = __( 'The maximum stack depth has been exceeded', 'amp' );
+						break;
+					case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_STATE_MISMATCH:
+						$message = __( 'Invalid or malformed JSON', 'amp' );
+						break;
+					// ...
+					default:
+						if ( isset( $validation_error['message'] ) ) {
+							$message = $validation_error['message'];
+						}
+				}
+				if ( $message ) {
+					$content .= sprintf( '<p>%s</p>', esc_html( $message ) );
 				}
 
 				break;
@@ -3099,6 +3113,11 @@ class AMP_Validation_Error_Taxonomy {
 				}
 				return $title;
 
+			case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_DEPTH:
+			case AMP_Tag_And_Attribute_Sanitizer::JSON_ERROR_STATE_MISMATCH:
+			// ...
+				return __( 'Invalid JSON', 'amp' );
+
 			default:
 				/* translators: %s error code */
 				return sprintf( __( 'Unknown error (%s)', 'amp' ), $validation_error['code'] );

```

Reply:
Thanks, how does https://github.com/ampproject/amp-wp/pull/4340/commits/47565f19398e500141b1ea839804c1de1a530268 look?

`JSON_ERROR_EMPTY` isn't one of the [json_last_error()](https://www.php.net/manual/en/function.json-last-error.php#refsect1-function.json-last-error-returnvalues) values, but I made it up for `<script>` textContent of `''` or `' '`.

If it's alright, later today I'll add a test for `AMP_Validation_Error_Taxonomy::filter_manage_custom_columns()` that covers that new `switch` statement for the message.

Reply:
https://github.com/ampproject/amp-wp/pull/4340/commits/47b8e8706e56a634bbe745701765f1bb4762ad24 adds a test for the error message.